### PR TITLE
AIP-84 Get Task Instance Try Details

### PR DESCRIPTION
**Pull request by kandharvishnu**
related to: https://github.com/apache/airflow/issues/42370

---

### CODE REVIEW

File: tests/api_fastapi/core_api/routes/public/test_task_instances.py [hunk 1]
```diff
@@ -143,20 +143,6 @@ def create_task_instances(
             session.commit()
         return tis
 
-        session.commit()
```
Comment (kandharvishnu) [inline tests/api_fastapi/core_api/routes/public/test_task_instances.py:146]: This code appears to be repetitive, so I’m removing it.

File: tests/api_fastapi/core_api/routes/public/test_task_instances.py [hunk 2]
```diff
@@ -465,3 +451,55 @@ def test_should_respond_404_wrong_map_index(self, test_client, session):
...
+            "try_number": 1,
+            "unixname": getuser(),
+            "dag_run_id": "TEST_DAG_RUN_ID",
+            "rendered_fields": {},
+            "rendered_map_index": None,
+            "trigger": None,
+            "triggerer_job": None,
+        }
+
+    def test_should_respond_404_wrong_task_instance_try_details(self, test_client, session):
```
Comment (rawwar) [inline tests/api_fastapi/core_api/routes/public/test_task_instances.py]: Can you also add a test for "Multiple Task Instances" found error?  If this case is not even possible to occur, maybe its not required?

@pierrejeambrun , wdyt?

Comment (rawwar) [reply tests/api_fastapi/core_api/routes/public/test_task_instances.py]: Or refer to legacy tests [here](https://github.com/apache/airflow/blob/c559b208949832d81c0a823ace04d3e172f0a842/tests/api_connexion/endpoints/test_task_instance_endpoint.py#L2609)

File: airflow/api_fastapi/core_api/routes/public/task_instances.py [hunk 3]
```diff
@@ -335,3 +337,37 @@ def get_task_instances(
         ],
         total_entries=total_entries,
     )
+
+
+@task_instances_router.get(
+    "/{task_id}/tries/{task_try_number}", responses=create_openapi_http_exception_doc([401, 403, 404])
```
Comment (rawwar) [inline airflow/api_fastapi/core_api/routes/public/task_instances.py]: 401 and 403 are now added to the base router. Refer to https://github.com/apache/airflow/pull/43932
```suggestion
    "/{task_id}/tries/{task_try_number}", responses=create_openapi_http_exception_doc([status.HTTP_404_NOT_FOUND])
```

File: airflow/api_fastapi/core_api/routes/public/task_instances.py [hunk 4]
```diff
@@ -335,3 +337,40 @@ def get_task_instances(
...
+        task_instance = session.scalar(query)
+        return task_instance
+
+    result = _query(TI) or _query(TIH)
+    if result is None:
+        raise HTTPException(
+            status.HTTP_404_NOT_FOUND,
+            f"The Task Instance with dag_id: `{dag_id}`, run_id: `{dag_run_id}`, task_id: `{task_id}`, try_number: `{task_try_number}` and map_index: `{map_index}` was not found",
+        )
+    return TaskInstanceHistoryResponse.model_validate(result, from_attributes=True)
```
Comment (kandharvishnu) [inline airflow/api_fastapi/core_api/routes/public/task_instances.py:450]: @pierrejeambrun @ephraimbuddy shall we return `TaskInstanceHistoryResponse` or change it to `TaskInstanceResponse`

[related PR](https://github.com/apache/airflow/pull/43830)

File: airflow/api_fastapi/core_api/routes/public/task_instances.py [hunk 5]
```diff
@@ -335,3 +337,40 @@ def get_task_instances(
         ],
         total_entries=total_entries,
     )
+
+
+@task_instances_router.get(
+    "/{task_id}/tries/{task_try_number}",
+    responses=create_openapi_http_exception_doc(
+        [status.HTTP_401_UNAUTHORIZED, status.HTTP_403_FORBIDDEN, status.HTTP_404_NOT_FOUND]
```
Comment (rawwar) [inline airflow/api_fastapi/core_api/routes/public/task_instances.py]:
```suggestion
        [status.HTTP_404_NOT_FOUND]
```

Comment (rawwar) [reply airflow/api_fastapi/core_api/routes/public/task_instances.py]: we don't need to add 401, 403 separately. They are now automatically added

Comment (kandharvishnu) [reply airflow/api_fastapi/core_api/routes/public/task_instances.py]: oh, got it

File: airflow/api_fastapi/core_api/datamodels/task_instances.py [hunk 6]
```diff
@@ -83,3 +83,39 @@ class TaskDependencyCollectionResponse(BaseModel):
     """Task scheduling dependencies collection serializer for responses."""
 
     dependencies: list[TaskDependencyResponse]
+
+
+class TaskInstanceHistoryResponse(BaseModel):
+    """TaskInstance serializer for responses."""
```
Comment (rawwar) [inline airflow/api_fastapi/core_api/datamodels/task_instances.py]:
```suggestion
    """TaskInstanceHistory serializer for responses."""
```

File: airflow/api_fastapi/core_api/routes/public/task_instances.py (hunk 4 above)
Comment (pierrejeambrun) [reply airflow/api_fastapi/core_api/routes/public/task_instances.py:450]: As we saw that there was some small differences bestween `TaskInstanceHistoryResponse` and `TaskInstanceResponse`, I would say that `TaskInstanceHistoryResponse` is the most appropriate. (Those are not fully exchangeable)

**Comment by pierrejeambrun** [comment]
Need rebasing as well.

File: airflow/api_fastapi/core_api/datamodels/task_instances.py [hunk 7]
```diff
@@ -83,3 +83,39 @@ class TaskDependencyCollectionResponse(BaseModel):
...
+    priority_weight: int | None
+    operator: str | None
+    queued_dttm: datetime | None = Field(alias="queued_when")
+    pid: int | None
+    executor: str | None
+    executor_config: Annotated[str, BeforeValidator(str)]
+
+
+class TaskInstanceHistoryCollectionResponse(BaseModel):
+    """Task Instance Collection serializer for responses."""
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/datamodels/task_instances.py]: docstring need update.

File: airflow/api_fastapi/core_api/routes/public/task_instances.py [hunk 8]
```diff
@@ -331,3 +333,38 @@ def get_task_instances(
...
+    dag_id: str,
+    dag_run_id: str,
+    task_id: str,
+    task_try_number: int,
+    session: Annotated[Session, Depends(get_session)],
+    map_index: int = -1,
+) -> TaskInstanceHistoryResponse:
+    """Get task instance details by try number."""
+
+    def _query(TI):
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/task_instances.py]:
```suggestion
    def _query(TI):
```
```suggestion
    def _query(orm_object: Base):
```
`orm_object` is more explicit.

Comment (pierrejeambrun) [reply airflow/api_fastapi/core_api/routes/public/task_instances.py]: Also you can add typing (param and return type) Select

File: tests/api_fastapi/core_api/routes/public/test_task_instances.py [hunk 9]
```diff
@@ -1127,3 +1128,174 @@ def test_should_respond_dependencies_mapped(self, test_client, session):
...
+            "queued_when": None,
+            "start_date": "2020-01-02T00:00:00Z",
+            "state": "success",
+            "task_id": "print_the_context",
+            "task_display_name": "print_the_context",
+            "try_number": 1,
+            "unixname": getuser(),
+            "dag_run_id": "TEST_DAG_RUN_ID",
+        }
+
```
Comment (pierrejeambrun) [inline tests/api_fastapi/core_api/routes/public/test_task_instances.py:1444]: Any reason why `test_should_respond_200_with_mapped_task_at_different_try_numbers` was not moved over ?

**Review by pierrejeambrun — COMMENTED** [review]
Overall looking good, a few suggestions, thanks

File: tests/api_fastapi/core_api/routes/public/test_task_instances.py (hunk 9 above)
Comment (kandharvishnu) [reply tests/api_fastapi/core_api/routes/public/test_task_instances.py:1444]: This method needs to be added after the implementation of `get_mapped_task_instance_try_details` 

https://github.com/apache/airflow/blob/fd0a0d2eb2f24452e67f25a3de1eb6e27a901203/tests/api_connexion/endpoints/test_task_instance_endpoint.py#L2708

Comment (pierrejeambrun) [reply tests/api_fastapi/core_api/routes/public/test_task_instances.py:1444]: Oh, that's a different endpoint for mapped task, got it.

**Comment by pierrejeambrun** [comment]
Needs rebasing. We can merge after the small adjustments are done.

**Review by pierrejeambrun — APPROVED** [review]

**Comment by rawwar** [comment]
Nice! On to the next one @kandharvishnu

**Pull request merged**